### PR TITLE
http_listener: pre-allocate buffer for reading request body

### PR DESCRIPTION
**Pull request by supershal**
### Required for all PRs:

- [ ] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [X] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [ ] README.md updated (if adding a new plugin)

Fixes: #1823 and #1856
We are using http_listener as a relay to route traffic between two influxdb instances in our HA setup. We noticed memory leaks when sending avg 800 metrics per batch from around 500 instances running telegraf. The bugfix #1826 caused issue with metrics parsing #1856.  

This fix handles the incoming requests the same way as influxdb's HTTP write handler. It reads content-length header and preallocates buffer with that size and reads all data before parsing them using scanner. We have done some stress tests on the fix and running the fixed version since two days without any issue. The telegraf instances does not seems to leak memory and memory usage remains low.

### CODE REVIEW

**Comment by sparrc** [comment]
@supershal do you happen to have any steps to reproduce this? I haven't been able to under a very high load using our influx stress tool (>300,000 points/s)

I think that this is going to have the same issue that #1826 was intending to solve, which is that we are basically doing an ioutil.ReadAll with the request body, from what I can tell of this change.

we will need to test with very large request bodies, larger than 100,000 points.

**Comment by supershal** [comment]
@sparrc I notice higher memory usage when http_listener timeouts while reading the request body. 
With old code (with ioutil.ReadAll), I was able to reproduce issue with our staging environment by reducing http_listener's config "read_timeout" to less than 10s and 100 http requests/second with 1000 points each. 

With fix above, when I used the same timeout for the same traffic, I do see a lot of timeout errors in log, but memory is not going up.

 We are using c3.large instances in our staging environment. 
I was not able to reproduce it on my local test. I suspect GC is collecting memory faster on my local box. I have to create similar configuration as c3.large (2 core) in order to consistently reproduce it. I will try to create comprehensive stress test tonight. 

I think the difference between ioutil.ReadAll and above fix is the pre-allocating buffer capacity so the buffer byte slice does not have to expand frequently while reading bytes from request body. I think that could be reason for timeouts. 
ioutil.ReadAll uses 512 bytes initially. https://golang.org/src/io/ioutil/ioutil.go?s=1147:1188#L31
Buffer.ReadFrom expands it once it reach it capacity. https://golang.org/src/bytes/buffer.go?s=5609:5668#L147
If we have enough capacity allocated in advance for the buffer, rapid expansion of buffer and copy of the contents would not trigger frequently. 
However I dont know how GC is claiming back memory in both case. I did not noticed memory going up with above fix and introducing timeouts.

**Comment by sparrc** [comment]
it makes some sense that pre-allocating the buffer rather than expanding it starting with a 512-byte buffer would perform better, though I'm not sure I understand why the memory usage is so much lower either.

`bytes.Buffer` also has an incremental reading function `Buffer.ReadBytes(delim)` that could be used to read & parse the buffer line-by-line.

We may want to look into using that as well. I know that in InfluxDB it's much more efficient to batch and write all the points together, and this might be why they didn't do it incrementally.

**Comment by sparrc** [comment]
@supershal I've been experiencing much better performance and more consistent writes using your changes 👍 . I've made a few adjustments, mainly:

1. limit the size of the request using http.MaxBytesReader
1. use the http.Request object's ContentLength value rather than parsing the header using strconv
1. continue through the function if there is a failure in the parser.Parse function, because it's possible for this function to fail but still have parsed _some_ metrics correctly.

**Comment by supershal** [comment]
@sparrc Thanks for making enhancements to the fix.

**Comment by supershal** [comment]
@sparrc These changes helps a lot to read http body faster. 
However request processing through telegraf is not fast enought to use http_listener + telegraf as relay between telegrafs and influxdb.
I quickly hacked up this brute force program to test standalone influxdb and http_listener + telegraf. https://gist.github.com/supershal/706adee2062817d0476df18af4b1e166
It generates similar load as our production environment. 
I tested it directly against local influxdb Vs http_listerner -> telegraf -> file, influxdb.  
Direct influxdb inserts for the same traffic takes on avg 500 ms while http_listener + telegraf takes more than 10s. This causes a lot of timeouts and http requests are stacked up.

While going through code, I noticed that metrics get flushed to output writer when metricbatchsize is reached while adding a point to each output buffer.
https://github.com/influxdata/telegraf/blob/master/internal/models/running_output.go#L76
It gets worse when we have multiple output plugins since points added to each output serially. https://github.com/influxdata/telegraf/blob/master/agent/agent.go#L280
I think this is holding the channel in Accumulator and eventually the input plugin.

I will play around with couple of fixes to see if I see as similar performance as against single influxdb. Let me know what are your thoughts and suggestion to improve the performance.

**Comment by sparrc** [comment]
@supershal I've made some changes to your original PR,

I don't think that reading the entire http body as fast as possible is the absolute best solution, because in some instances the body might be very large and we don't want to create very large allocations. For that reason I've changed it to read the body byte-by-byte up to a maximum 10MB buffer, then flush each time that fills.

Any perf improvement PRs are appreciated but I wouldn't recommend relying on Telegraf as a high-throughput relay to InfluxDB and I wouldn't recommend that you implement this setup in you environment, regardless of the performance improvements we can make.

From what you've said so far I'm fairly certain that simply using the `flush_jitter` config option in telegraf you will have no problems writing directly to influxdb, this is exactly the use-case that config option was designed for.

File: plugins/inputs/http_listener/http_listener.go [hunk 1]
```diff
@@ -111,36 +133,112 @@ func (t *HttpListener) httpListen() error {
...
+					}
+					break
+				}
+				// returned error is always nil:
+				// https://golang.org/pkg/bytes/#Buffer.WriteByte
+				buffer.WriteByte(b)
+				// if we have a newline and we're nearing the end of the buffer,
+				// do a write and continue with a fresh buffer.
+				if buffer.Len() > MAX_ALLOCATION_SIZE-256*1000 && b == '\n' {
+					t.parse(buffer.Bytes(), &msg400)
```
Comment (joelegasse) [inline plugins/inputs/http_listener/http_listener.go]: This chunked processing differs in behavior from the standard influxdb handler in that these "chunks" will end up with separate timestamps, if they weren't set on the original points. If there's going to be chunked processing, the current time should be grabbed at the beginning of the processing, then passed around as the default time to `models.ParsePointsWithPrecision`. I see that `models.ParsePoints` is used in the parser, which just [calls](https://github.com/influxdata/influxdb/blob/8fce6bba4814df34017117ba97270791520936af/models/points.go#L178) `models.ParsePointsWithPrecision(buf, time.Now(), "n")`. Capturing the timestamp earlier and replacing all uses of `ParsePoints(buf)` with `ParsePointsWithPrecision(buf, ts, "n")` would keep similar behavior to the database handler.

**Comment by sparrc** [comment]
thank you for the contribution @supershal, I'm going to close this in favor of https://github.com/influxdata/telegraf/pull/1915

Our hope is that #1915 will perform better in the long run than reading the request body all at once.  It's possible that in some situations it may seem as if the reading of the request body is timing out more, because it does iteratively read the body.

That being said, it should allocate significantly less memory in regular average usage, and should also put substantially less pressure on GC.

**Pull request closed without merging**